File: client/src/main/java/com/chatroom/client/ChatClient.java
package com.chatroom.client;

import com.chatroom.shared.MessageProtocol;

import java.io.*;
import java.net.Socket;
import java.util.logging.Logger;

public class ChatClient {
    private static final Logger LOGGER = Logger.getLogger(ChatClient.class.getName());

    private final ClientListener listener;
    private volatile Socket socket;
    private volatile ObjectOutputStream output;
    private volatile ObjectInputStream input;
    private volatile boolean connected = false;
    private volatile MessageListener messageListener;
    private volatile Thread listenerThread;
    private String userName = "Anonymous";

    public ChatClient(ClientListener listener) {
        this.listener = listener;
    }

    public void connect(String host, int port, String username) {
        if (connected) {
            return;
        }
        this.userName = username;

        try {
            socket = new Socket(host, port);
            output = new ObjectOutputStream(socket.getOutputStream());
            output.flush();
            input = new ObjectInputStream(socket.getInputStream());
            input.setObjectInputFilter(
                java.io.ObjectInputFilter.Config.createFilter("java.lang.String;!*"));

            // Send username
            output.writeObject(userName);
            output.flush();

            connected = true;
            messageListener = new MessageListener(input, listener, this);
            listenerThread = new Thread(messageListener, "MessageListener-" + userName);
            listenerThread.start();
        } catch (IOException e) {
            connected = false;
            listener.onConnectionError("Could not connect: " + e.getMessage());
        }
    }

    public void send(String message) {
        if (!connected || output == null) {
            return;
        }

        try {
            String writeStr;
            if (message.startsWith(MessageProtocol.PREFIX_PRIVATE)) {
                listener.onMessageReceived(userName + ": " + message);
                writeStr = message;
            } else {
                writeStr = MessageProtocol.PREFIX_BROADCAST + userName + ": " + message;
            }
            synchronized (output) {
                output.writeObject(writeStr);
                output.flush();
            }
        } catch (IOException e) {
            listener.onConnectionError("Failed to send message");
        }
    }

    public void disconnect() {
        connected = false;
        if (messageListener != null) {
            messageListener.stop();
        }
        closeQuietly();
    }

    void onListenerDisconnect() {
        connected = false;
        closeQuietly();
    }

    private void closeQuietly() {
        try { if (output != null) output.close(); } catch (IOException e) {}
        try { if (input != null) input.close(); } catch (IOException e) {}
        try { if (socket != null && !socket.isClosed()) socket.close(); } catch (IOException e) {}
    }
}

File: client/src/main/kotlin/com/chatroom/client/ChatClient.kt
package com.chatroom.client

import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.Socket
import kotlin.concurrent.thread

class ChatClient(private val listener: ClientListener) {
    private var socket: Socket? = null
    private var output: ObjectOutputStream? = null
    private var input: ObjectInputStream? = null
    @Volatile private var running = false
    var userName: String = "Anonymous"

    fun connect(host: String, port: Int, username: String) {
        this.userName = username
        thread {
            try {
                socket = Socket(host, port)
                output = ObjectOutputStream(socket?.getOutputStream())
                output?.flush()
                input = ObjectInputStream(socket?.getInputStream())

                // Send username
                output?.writeObject(username)
                output?.flush()

                running = true
                startListening()
            } catch (e: Exception) {
                listener.onConnectionError("Could not connect: ${e.message}")
            }
        }
    }

    private fun startListening() {
        try {
            while (running) {
                val messageObj = input?.readObject() ?: break
                val message = messageObj.toString()

                if (message.startsWith("!")) {
                    // User list update: ![User1, User2]
                    val usersStr = message.substring(1) // [User1, User2]
                    val users = usersStr.removePrefix("[").removeSuffix("]").split(", ")
                        .filter { it.isNotBlank() }
                    listener.onUserListUpdated(users)
                } else {
                     if (message.startsWith("@EE@|")) {
                         listener.onMessageReceived(message.substring(5))
                     } else if (message.startsWith("@")) {
                         listener.onMessageReceived(message.substring(1))
                     } else {
                         // Fallback
                         listener.onMessageReceived(message)
                     }
                }
            }
        } catch (e: Exception) {
            if (running) {
                listener.onConnectionError("Connection lost")
            }
        } finally {
            disconnect()
        }
    }

    fun send(message: String) {
        if (!running) return

        try {
            val writeStr = if (message.startsWith("@")) {
                 listener.onMessageReceived("$userName: $message")
                 message
            } else {
                 "@EE@|$userName: $message"
            }

            output?.writeObject(writeStr)
            output?.flush()
        } catch (e: Exception) {
            listener.onConnectionError("Failed to send message")
        }
    }

    fun disconnect() {
        running = false
        try {
            output?.close()
            input?.close()
            socket?.close()
        } catch (e: Exception) {}
    }
}
